ZEP000-wf-stream-processing.md

```diff
@@ -0,0 +1,858 @@
+> this time not yet clear whether we need the `Event_Occur` command and/or the
+> `Event_Occurred` event.
+
+> Also note, that `Sequence_Flow_Taken` is an event without a command. We'll
+> need to find a way to make this fit the strict commands/events separation.
+
+## Processing and Replay
+
+The following diagram shows from left to right a log stream consisting of
+command and event records. Command records can only be processed, while event
+records can only be replayed.
+
+![event-sourcing](ZEP-wf-stream-processing/event-sourcing.png)
+
+> The black arrow indicates that particular events and commands are appended to
+> the log stream as a result of processing a previous command (i.e. it
+> represents the source record position - explained later in more detail). The
+> orange box wrapping some records indicates that these were appended to log
+> stream atomically.
+
+### Activities during Processing
+
+The **commands** are processed only on processing and drive the workflow execution. The processing of a command includes:
+* read the state of the workflow instance and related resources (RocksDB)
+* check if the record can be applied (i.e. concurrent update guard)
+* apply the business logic  
+* write updates to the state
+* write events and (optionally) follow-up commands to continue the processing
+* apply side-effects (e.g. send a response)
+
+The following diagram shows the command processing loop.
+
+![command-processing](ZEP-wf-stream-processing/event-sourcing-processing.png)
+
+### Activities during Replay
+
+The **events** are processed only on replay and rebuild the state. The replay of an event includes:
+* write updates to the state
+
+The following diagram shows the event replay.
+
+![event-replay](ZEP-wf-stream-processing/event-sourcing-replay.png)
+
+In order to rebuild the state with the events, it is necessary that all state changes from the command processing can be applied with the events only.
+So, the command processing may only update the state when it also writes an event that reflects the state changes.
+
+When the replay of the events is completed, the resulting state is exactly the same as when processing stopped the last time.
+Therefore, the processing of the commands can start again now.
+
+## The boundary between replay and processing
+
+On restart, it replays all events of the stream, but ignores events that have a
+source record position that is less than or equal to the last processed command
+of the snapshot. The state changes of these events are already included in the
+snapshot, because the commands are processed before the snapshot was created.
+The replay ends when all events have been either replayed or ignored.
+
+When the replay is done, then the processing starts on the first command that
+has not yet been processed. A command is only considered to be processed if the
+stream contains a record (a command, an event, or a rejection) with a source
```

Comment:
> why we make a distinction between events and rejections?

Events and rejections have different semantic. A consumer of the stream can see directly if a command was processed or rejected. 

On processing, we can skip rejections. So, we can save processing power.

Does this answer your question?

Reply:
Yes, that's what I meant by the practical reason :slightly_smiling_face: Conceptually, it seems to me rejections are just a special type of events, but I asked because, maybe not :sweat_smile:

Reply:
Kind of. But the semantics are still different. An event says that something happened, for example, "the service task is activated". A rejection says that a command could not be applied. But nothing happened in this case.